cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,312 @@
+
+    CUDA_HOST_DEVICE_CALLABLE
+    T_output at(gdf_index_type id) const {
+        return T_output(get_value(id), is_valid(id));
+    };
+
+    CUDA_HOST_DEVICE_CALLABLE
+    T_element get_value(gdf_index_type id) const {
+        return (is_valid(id))? data[id] : identity;
+    }
+
+    CUDA_HOST_DEVICE_CALLABLE
+    bool is_valid(gdf_index_type id) const
+    {
+        return bit_mask::is_valid(valid, id);
+    }
+};
+
+
+// ---------------------------------------------------------------------------
+// column_input_iterator
+template<typename T_output, typename T_input, typename Iterator=thrust::counting_iterator<gdf_index_type> >
+  class column_input_iterator
+    : public thrust::iterator_adaptor<
+        column_input_iterator<T_output, T_input, Iterator>, // the first template parameter is the name of the iterator we're creating
+        Iterator,                   // the second template parameter is the name of the iterator we're adapting
+        thrust::use_default, thrust::use_default, thrust::use_default,
+        T_output,                   // set `super_t::reference` to `T_output`
+        thrust::use_default
+      >
+  {
+  public:
+    // shorthand for the name of the iterator_adaptor we're deriving from
+    using super_t = thrust::iterator_adaptor<
+      column_input_iterator<T_output, T_input, Iterator>, Iterator,
+      thrust::use_default, thrust::use_default, thrust::use_default, T_output, thrust::use_default
+    >;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_iterator(const T_input col, const Iterator &it) : super_t(it), colData(col){}
+
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_iterator(const T_input col) : super_t(Iterator{0}), colData(col){}
+
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_iterator(const column_input_iterator &it) : super_t(it.base()), colData(it.colData){}
+
+    // befriend thrust::iterator_core_access to allow it access to the private interface below
+    friend class thrust::iterator_core_access;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_iterator &operator=(const column_input_iterator &other)
+    {
+        super_t::operator=(other);
+
+        colData = other.colData;
+        return *this;
+    }
+
+  private:
+    const T_input colData;
+
+    // it is private because only thrust::iterator_core_access needs access to it
+    CUDA_HOST_DEVICE_CALLABLE
+    typename super_t::reference dereference() const
+    {
+      int id = *(this->base());
+      return colData.at(id);
+    }
+};
+
+// ---------------------------------------------------------------------------
+// helper functions to make iterator
+
+template <typename T_element, typename T_output = T_element,
+    typename T_output_helper = ColumnOutputSingle<T_output>,
+    typename Iterator_Index=thrust::counting_iterator<gdf_index_type> >
+auto make_iterator_with_nulls(const T_element *_data, const bit_mask::bit_mask_t *_valid,
```

Comment:
What about factories that take columns? 
```
auto make_iterator(gdf_column const& column) {...
```

Reply:
The iterator class with and without nulls must be separated at compile time.
Taking `gdf_column` is possible, but we need a pair of `make_iterator_with_nulls` and `make_iterator_without_nulls` for `gdf_column` input.
If it is okay, I will add a pair of factories from `gdf_column`.

Reply:
I've added a pair of `make_iterator_with_nulls` and `make_iterator_without_nulls` for `gdf_column `input.
Please check it.
Tomorrow I will add test cases for `make_iterator_XXX` with gdf_column input.

Reply:
Soon(ish) we will have a `cudf::column` class. Presumably these iterators will just be members. Will that not work if it's not known at compile time whether the column has nulls?

Reply:
@harrism